HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import division
+import numpy as np
+import pyspark
+
+def bfsSearch(startName, charAdj, sc):
+
+	currentSet = sc.parallelize([startName])
+	currentSet = currentSet.map(lambda x: (x,(None,float('inf')))).partitionBy(10).cache()
+	characterList = charAdj.map(lambda x: x[0])
+	counts = []
+	count = 0
+
+	bfsDistances = characterList.map(lambda x: (x,(None,float('inf')))).partitionBy(10).cache()
+
+	nNeighbors = sc.accumulator(0)
+
+	def updateDist(node):
+		if node[1][1] != None: # value present
+			if node[1][1][1] == float('inf'):
+				nNeighbors.add(1)
+			return (node[0],(node[1][1][0],min([node[1][0][1],count])))
+		else:
+			return (node[0],node[1][0])
+
+	oldValue = 0
+	newValue = 1
+
+	while newValue != oldValue:
+		bfsDistances = bfsDistances.leftOuterJoin(currentSet).cache() # (character,((storedPrevious,storedDist),(setDistPrevious,currentSetDist))), if currentSetDist = none, do nothing 
+		bfsDistances.count() 													#       else if new neighbor ('inf'), replace with count
+		oldValue = nNeighbors.value
+		bfsDistances = bfsDistances.map(updateDist).cache() # update distances (character,(prevNodes,updatedDistance))
+		bfsDistances.count()
```

Comment:
you could use foreach() loop to update the accumulator so you don't have to count(). But I also like hitting 2 birds with 1 stone.